### ISSUE TITLE
Refactor sorting to use guessit

### ISSUE THREAD

**Issue by Safihre** [sabnzbd/sabnzbd#1881]
Sorting has always been a mystery part of the code for me, it sort of just works.
Looking at NZBGet's VideoSort script it uses the `guessit` Python module.
It seems to be a actively maintained project that can do so much more than we currently can do with our own regex'es.

I propose to refactor our Sorting code to use `guessit`, which would also allow us to introduce even more variables for users to use in renaming.

**Comment by sanderjo** [comment]
I only use one sort of sorting: on the front page, like Sort by Size.

I think you're talking about the other 'Sorting', on http://127.0.0.1:8080/config/sorting/
So that is Sorting (or: Storing) in subdirectories based on the file name, right? Is this a right moment to suggest renaming to Storing? Or at least put in a title "Sorting into subdirectories"? Even the https://sabnzbd.org/wiki/configuration/3.2/sorting is unclear about that, IMHO.

The guessit module looks powerful. It would getting rid of https://github.com/sabnzbd/sabnzbd/blob/develop/sabnzbd/sorting.py completely? That would be nice.

From https://pypi.org/project/guessit/

GuessIt is a python library that extracts as much information as possible from a video filename.


```
$ guessit "Treme.1x03.Right.Place,.Wrong.Time.HDTV.XviD-NoTV.avi"
For: Treme.1x03.Right.Place,.Wrong.Time.HDTV.XviD-NoTV.avi
GuessIt found: {
    "title": "Treme",
    "season": 1,
    "episode": 3,
    "episode_title": "Right Place, Wrong Time",
    "source": "HDTV",
    "video_codec": "Xvid",
    "release_group": "NoTV",
    "container": "avi",
    "mimetype": "video/x-msvideo",
    "type": "episode"
}
```

**Comment by jcfp** [comment]
> I propose to refactor our Sorting code to use `guessit`, which would also allow us to introduce even more variables for users to use in renaming.

Great idea. Probably could get rid of the different sorters in the backend altogether. Detecting type, date, languages, etc is almost too easy.

**Comment by Safihre** [comment]
We can then also use this for duplicate detection. For example I received a user-report today that these 2 filenames were not detected as duplicates:
```python
sabnzbd.newsunpack.analyse_show("Magnum.P.I.2018.S03E13.720p.HEVC.x265-MeGusta.nzb")
('Magnum P I 2018', '3', '13', '720p HEVC x265-MeGusta nzb')
sabnzbd.newsunpack.analyse_show("Magnum.P.I.S03E13.720p.HEVC.x265-MeGusta.nzb")
('Magnum P I', '3', '13', '720p HEVC x265-MeGusta nzb')
```

**Comment by Safihre** [comment]
@jcfp Did you make any progress on this? I did notice that when I put any random name in `guessit`, it always detects it as `movie`. So even things like `setup.exe`. So we can't rely fully on the type-detection. But maybe that's not even needed?

**Comment by jcfp** [comment]
Nothing ready to go yet. Guessit has more issues like that, for example with titles that start with digits (like the "24" tv series) or movies getting detected as a series in case there's no year in the jobname. I think most of these can be worked around, for example by hinting the type where possible, prepending a random string to the title, or requiring at least some common attributes are set for the movie type (e.g. at least n out of resolution, year, audio, video codec) to filter out the "setup.exe" movies.

**Comment by Safihre** [comment]
@jcfp If the current SABnzbd code doesn't handle it right, it's fine if Guessit doesn't handle it either 😉

**Comment by JackDandy** [comment]
> ```python
> sabnzbd.newsunpack.analyse_show("Magnum.P.I.2018.S03E13.720p.HEVC.x265-MeGusta.nzb")
> ('Magnum P I 2018', '3', '13', '720p HEVC x265-MeGusta nzb')
> sabnzbd.newsunpack.analyse_show("Magnum.P.I.S03E13.720p.HEVC.x265-MeGusta.nzb")
> ('Magnum P I', '3', '13', '720p HEVC x265-MeGusta nzb')
> ```

Magnum is particularly problematic for detecting dupes because there is no standard way to know the 2018 series from the original back in the day, well there was, but the distinction got fuddled by some releases - no surprises there. Also, using release title for dupe check is problematic in any case, for example, an item added now could fail, then a reattempt to add it (via API) in 30 mins trips dupe check which prevents action, some kind of "force" flag would be useful from the perspective of adding via API.

**Comment by Safihre** [comment]
@JackDandy only successful downloads are checked for dupe-check. Current jobs in the queue are ignore, which is another feature request that's been open for ages: #727.

**Comment by Safihre** [comment]
Closed by #1916
